File: java/solution/src/test/java/Webshop/UnitTests/OrderServiceTests.java
package Webshop.UnitTests;

import Webshop.Service.Orders.OrderRepository;
import Webshop.Service.Orders.OrderResult;
import Webshop.Service.Users.User;
import org.junit.jupiter.api.Test;
import Webshop.Service.Common.NotificationService;
import Webshop.Service.Orders.OrderResultCode;
import Webshop.Service.Orders.OrderService;
import Webshop.Service.Products.ProductService;
import Webshop.Service.Products.Product;
import Webshop.Service.Users.UserService;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

public class OrderServiceTests {
    @Test
    public void placeOrder_whenProductIsNotFound_shouldReturnProductNotFound() {
        ProductService productService = mock(ProductService.class);
        UserService userService = mock(UserService.class);
        NotificationService notificationService = mock(NotificationService.class);
        OrderRepository orderRepository = mock(OrderRepository.class);

        when(productService.getProduct("product1")).thenReturn(null); // Simulate that product does not exist

        OrderService orderService = new OrderService(orderRepository, productService, userService, notificationService);

        OrderResult orderResult = orderService.placeOrder("product1", 1, "user1");

        assertFalse(orderResult.isSuccess(), "Expected order to fail");
        assertEquals(OrderResultCode.ProductNotFound, orderResult.getResultCode());
    }

    @Test
    public void placeOrder_whenInventoryIsEmpty_shouldReturnOutOfStock() {
        ProductService productService = mock(ProductService.class);
        UserService userService = mock(UserService.class);
        NotificationService notificationService = mock(NotificationService.class);
        OrderRepository orderRepository = mock(OrderRepository.class);

        when(productService.getProduct("product1")).thenReturn(new Product("product1", 10)); // Make sure the product exists
        when(productService.getProductInventory("product1")).thenReturn(0); // Simulate that inventory is empty

        OrderService orderService = new OrderService(orderRepository, productService, userService, notificationService);

        OrderResult orderResult = orderService.placeOrder("product1", 1, "user1");

        assertFalse(orderResult.isSuccess(), "Expected order to fail");
        assertEquals(OrderResultCode.OutOfStock, orderResult.getResultCode());
    }

    @Test
    public void placeOrder_whenSuccessful_shouldUpdateInventory() {
        ProductService productService = mock(ProductService.class);
        UserService userService = mock(UserService.class);
        NotificationService notificationService = mock(NotificationService.class);
        OrderRepository orderRepository = mock(OrderRepository.class);

        when(userService.getUser("user1")).thenReturn(new User("user1", "a", "b", "c", "Abcd1234")); // Make sure the user exists
        when(productService.getProduct("product1")).thenReturn(new Product("product1", 10)); // Make sure the product exists
        when(productService.getProductInventory("product1")).thenReturn(4); // Simulate that inventory has enough items

        OrderService orderService = new OrderService(orderRepository, productService, userService, notificationService);

        OrderResult orderResult = orderService.placeOrder("product1", 2, "user1");

        assertTrue(orderResult.isSuccess(), "Expected order to be successful");
        verify(productService, times(1)).subtractFromProductInventory("product1", 2);
    }

    @Test
    public void placeOrder_whenSuccessful_shouldSendNotification() {
        ProductService productService = mock(ProductService.class);
        UserService userService = mock(UserService.class);
        NotificationService notificationService = mock(NotificationService.class);
        OrderRepository orderRepository = mock(OrderRepository.class);

        when(userService.getUser("user1")).thenReturn(new User("user1", "a", "b", "c", "Abcd1234")); // Make sure the user exists
        when(productService.getProduct("product1")).thenReturn(new Product("product1", 1)); // Make sure the product exists
        when(productService.getProductInventory("product1")).thenReturn(1); // Simulate that inventory has enough items

        OrderService orderService = new OrderService(orderRepository, productService, userService, notificationService);

        OrderResult orderResult = orderService.placeOrder("product1", 1, "user1");

        assertTrue(orderResult.isSuccess(), "Expected order to be successful");
        verify(notificationService, times(1)).sendMail(anyString(), anyString());
    }
}


// Run the test with the following command:
// mvn -Dtest=OrderServiceTests test

// Run a single test with the following command:
// mvn -Dtest=OrderServiceTests#placeOrder_whenProductIsNotFound_shouldReturnProductNotFound test

File: csharp/solution/Webshop.UnitTests/OrderServiceTests.cs
using FakeItEasy;
using NUnit.Framework;
using Webshop.Service.Common;
using Webshop.Service.Orders;
using Webshop.Service.Products;
using Webshop.Service.Users;

namespace Webshop.UnitTests;

[TestFixture]
public class OrderServiceTests
{
    [Test]
    public void PlaceOrder_WhenProductIsNotFound_ShouldReturnProductNotFound()
    {
        var orderRepository = A.Fake<IOrderRepository>();
        var productService = A.Fake<IProductService>();
        var userService = A.Fake<IUserService>();
        var notificationService = A.Fake<INotificationService>();

        A.CallTo(() => productService.GetProduct("product1")).Returns(null); // Simulate that product does not exist

        var orderService = new OrderService(orderRepository, productService, userService, notificationService);

        var orderResult = orderService.PlaceOrder("product1", 1, "user1");

        Assert.That(orderResult.IsSuccess, Is.False, "Expected order to fail");
        Assert.That(orderResult.ResultCode, Is.EqualTo(OrderResultCode.ProductNotFound));
    }

    [Test]
    public void PlaceOrder_WhenInventoryIsEmpty_ShouldReturnOutOfStock()
    {
        var orderRepository = A.Fake<IOrderRepository>();
        var productService = A.Fake<IProductService>();
        var userService = A.Fake<IUserService>();
        var notificationService = A.Fake<INotificationService>();

        A.CallTo(() => productService.GetProduct("product1")).Returns(new Product { Id = "product1" }); // Make sure the product exists
        A.CallTo(() => productService.GetProductInventory("product1")).Returns(0); // Simulate that inventory is empty

        var orderService = new OrderService(orderRepository, productService, userService, notificationService);

        var orderResult = orderService.PlaceOrder("product1", 1, "user1");

        Assert.That(orderResult.IsSuccess, Is.False, "Expected order to fail");
        Assert.That(orderResult.ResultCode, Is.EqualTo(OrderResultCode.OutOfStock));
    }

    [Test]
    public void PlaceOrder_WhenSuccessful_ShouldUpdateInventory()
    {
        var orderRepository = A.Fake<IOrderRepository>();
        var productService = A.Fake<IProductService>();
        var userService = A.Fake<IUserService>();
        var notificationService = A.Fake<INotificationService>();

        A.CallTo(() => productService.GetProduct("product1")).Returns(new Product { Id = "product1" }); // Make sure the product exists
        A.CallTo(() => productService.GetProductInventory("product1")).Returns(4); // Simulate that inventory has enough items

        var orderService = new OrderService(orderRepository, productService, userService, notificationService);

        var orderResult = orderService.PlaceOrder("product1", 2, "user1");

        Assert.That(orderResult.IsSuccess, Is.True, "Expected order to be successful");
        A.CallTo(() => productService.SubtractFromProductInventory("product1", 2)).MustHaveHappenedOnceExactly();
    }

    [Test]
    public void PlaceOrder_WhenSuccessful_ShouldSendNotification()
    {
        var orderRepository = A.Fake<IOrderRepository>();
        var productService = A.Fake<IProductService>();
        var userService = A.Fake<IUserService>();
        var notificationService = A.Fake<INotificationService>();

        A.CallTo(() => productService.GetProduct("product1")).Returns(new Product { Id = "product1" }); // Make sure the product exists
        A.CallTo(() => productService.GetProductInventory("product1")).Returns(1); // Simulate that inventory has enough items

        var orderService = new OrderService(orderRepository, productService, userService, notificationService);

        var orderResult = orderService.PlaceOrder("product1", 1, "user1");

        Assert.That(orderResult.IsSuccess, Is.True, "Expected order to be successful");
        A.CallTo(() => notificationService.SendMail(A<string>.Ignored, A<string>.Ignored)).MustHaveHappenedOnceExactly();
    }
}